# GSEA

### Setup environment

```{r}
library(DESeq2)
library(magrittr)
```

### Get ginger vs *A. thaliana* blastp results

Ginger protein -\> *A. thaliana* protein

```{r}
blastp_res <- read.table(
  file = "outputs/blastp/ginger_vs_ath.tsv",
  header = FALSE,
  sep = "\t"
)
head(blastp_res)

```

### Get protein ID -\> gene ID relationship for ginger and *A. thaliana*

#### Define GTF utils

Define parser for GTF attributes

```{r}
parse_attribute <- function(string) {
  tmp <- string %>%
    trimws(., which = "both") %>% # 去除字串開頭與/或結尾的空白字元
    strsplit(., ";") %>%
    unlist %>% # 攤平成向量
    trimws(., which = "both") %>% #再去掉頭尾空白
    strsplit(., " ") # 以空格切分成 key/value 向量
  out <- c()
  for (i in tmp) {
    out[i[1]] = paste(i[-1], collapse = " ") # i[-1]：除了第一個以外的所有元素組成一個字串向量，也就是value各段。
  }
  return(out)
}
```

GTF header

```{r}
gtf_header <- c( # 定義讀入 GTF 檔案時要用的欄位名稱
  "seqname",
  "source",
  "feature",
  "start",
  "end",
  "score",
  "strand",
  "frame",
  "attributes"
)
```

1.  Ginger

```{r}
# import gtf
ginger_gtf <- read.table(
  file = "references/ncbi_dataset/data/GCF_018446385.1/genomic.finished.gtf",
  sep = "\t",
  header = FALSE,
  col.names = gtf_header,
  comment.char = "#"
)

# find every first CDS lines of each protein coding gene
cds_lines <- ginger_gtf$feature == "CDS"
every_first_cds <- 
  (cds_lines & c(FALSE, !cds_lines[-length(cds_lines)])) %>% 
  which

# get protein ID and gene ID from parsed attributes
ginger_p2g <- matrix(
  ncol = 2,
  nrow = length(every_first_cds)
)
j <- 1
for (nth_row in every_first_cds) {
  attrs <- parse_attribute(ginger_gtf$attributes[nth_row])
  ginger_p2g[j, 1] <- attrs["protein_id"]
  ginger_p2g[j, 2] <- attrs["gene_id"]
  j <- j + 1
}

# set row name of the matrix as protein ID
rownames(ginger_p2g) <- ginger_p2g[, 1]
colnames(ginger_p2g) <- c("pid", "gid")
```